exercises/dnd-character/canonical-data.json

```diff
@@ -0,0 +1,176 @@
+{
+  "exercise": "dnd-character",
+  "version": "1.0.0",
+  "comments": [
+    "The random generator 'ability' can be property-tested.",
+    "The pure function 'modifier' can be tested totally.",
+    "The random generator 'character' should return some",
+    "kind of structure with six random abilities and a",
+    "hitpoint score. This structure can be property-tested",
+    "in the same way as 'ability' and 'modifier'.",
+    "",
+    "It may be fun to provide a 'name' property for characters",
+    "that may or may not be randomly generated but not tested.",
+    "",
+    "Many programming languages feature property-based test",
+    "frameworks. For a language-agnostic introduction to",
+    "property-based testing, see:",
+    "",
+    "https://hypothesis.works/articles/what-is-property-based-testing/",
+  ],
+  "cases": [
+    {
+      "description": "random ability is within range",
+      "property": "ability",
+      "input": {},
+      "expected": "score >= 3 && score <= 18"
+    },
+    {
+      "description": "random character is valid",
+      "property": "character",
+      "input": {},
+      "expected": {
+        "strength": "strength >= 3 && strength <= 18",
+        "dexterity": "dexterity >= 3 && dexterity <= 18",
+        "constitution": "constitution >= 3 && constitution <= 18",
+        "intelligence": "intelligence >= 3 && intelligence <= 18",
+        "wisdom": "wisdom >= 3 && wisdom <= 18",
+        "charisma": "charisma >= 3 && charisma <= 18",
+        "hitpoints": "hitpoints == 10 + modifier(constitution)"
+      }
+    },
+    {
+      "description": "ability modifier",
```

Comment:
I was thinking that perhaps the "ability modifier" tests should be listed before the "ability" and "character" tests. I have two reasons for this:

- It is easier for students to start with implementing a "regular" function.
- The "character" test builds on "modifier" property, and it makes sense to introduce dependencies before using them.

Reply:
OK, good argument. If you think that the pure `modifier` function should be implemented before the two random generators `ability` and `character`, I'll list them first. It's just that the tests for `ability` seem to flood canonical-data.json.

The current order is based on the same reasoning as you apply wrt. dependencies.